samples/07.advanced-web-chat-apps/e.sso-on-behalf-of-authentication/app/src/oauth/Composer.js

```diff
@@ -0,0 +1,148 @@
+import PropTypes from 'prop-types';
+import React, { useCallback, useEffect, useMemo, useState } from 'react';
+import { UserAgentApplication } from 'msal';
+
+import Context from './Context';
+import fetchProfileDisplayName from './fetchProfileDisplayName';
+import fetchProfilePhotoAsBase64 from './fetchProfilePhotoAsBase64';
+import requiresInteraction from '../utils/requiresInteraction';
+import fetchJSON from '../utils/fetchJSON';
+
+const GRAPH_SCOPES = {
+  OPENID: 'openid',
+  PROFILE: 'profile',
+  USER_READ: 'User.Read',
+  MAIL_READ: 'Mail.Read'
+};
+
+const GRAPH_REQUESTS = {
+  LOGIN: {
+    scopes: [GRAPH_SCOPES.OPENID, GRAPH_SCOPES.PROFILE, GRAPH_SCOPES.USER_READ]
+  },
+  EMAIL: {
+    scopes: [GRAPH_SCOPES.MAIL_READ]
+  }
+};
+
+// Composer will prepare a React context object to use by consumer.
+const Composer = ({
+  children,
+  onError = error => {
+    console.error(error);
+  }
+}) => {
+  const [accessToken, setAccessToken] = useState();
+  const [avatarURL, setAvatarURL] = useState();
+  const [msal, setMsal] = useState();
+  const [name, setName] = useState();
+
+  const acquireToken = useCallback(
+    async (request = GRAPH_REQUESTS.LOGIN) => {
+      if (msal) {
+        return msal.acquireTokenSilent(request).catch(error => {
+          if (requiresInteraction(error)) {
+            return msal.acquireTokenPopup(request);
+          }
+        });
+      }
+    },
+    [msal]
+  );
+
+  const getAccount = useCallback(() => (msal ? msal.getAccount() : undefined), [msal]);
+
+  const onSignIn = useMemo(() => {
```

Comment:
Should we set a busy flag?

Since this is async, we should not allow bot/user to repeat sign-in while the process is ongoing.

If repeated attempt, should we silently ignore or throw exception?

Reply:
Yea that's probably a good idea. I feel like we should silently ignore the error.